### PR TITLE
[Reviewer: Seb] Split Astaire into two components:

### PR DESCRIPTION
**Pull request by richardwhiuk**
Seb,

I've split astaire into two:

 - Astaire, responsible for resyncing memcached instances, which is throttled as currently.
 - Rogers, responsible for proxying distributed memcached requests, which isn't throttled.

I've done the expected refactoring - moving root to astaire.root, creating rogers.root, adding the necessary Debain packaging, and Make packging

I've created a new source file, proxy_main.cpp which is the root file for rogers.

### CODE REVIEW

**Comment by richardwhiuk** [comment]
I've run the live tests across a multi-site deployment deployed with this change, and it looks good.

File: debian/control [hunk 1]
```diff
@@ -33,3 +33,29 @@ Section: debug
...
+Package: rogers-libs
+Architecture: any
+Description: Libraries for Rogers, memcached proxy server
+
+Package: rogers-libs-dbg
+Architecture: any
+Section: debug
+Priority: extra
+Depends: rogers-libs (= ${binary:Version}), gdb
+Description: Debugging symbols for Roger's libraries
```
Comment (sebrexmetaswitch) [inline debian/control]: Should be `Rogers'` :)

File: debian/rogers.init.d [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+# Metaswitch Networks in a separate written agreement.
+
+### BEGIN INIT INFO
+# Provides:          rogers
+# Required-Start:    $remote_fs $syslog
+# Required-Stop:     $remote_fs $syslog
+# Default-Start:     2 3 4 5
+# Default-Stop:      0 1 6
+# Short-Description: Astaire
+# Description:       Astaire, active resynchronisation for memcached clusters
```
Comment (sebrexmetaswitch) [inline debian/rogers.init.d]: The description should refer to Rogers, not Astaire

File: include/astaire_pd_definitions.hpp [hunk 3]
```diff
@@ -105,4 +105,47 @@ const static PDLog CL_ASTAIRE_COMPLETE_RESYNC
...
+
+const static PDLog CL_ROGERS_ENDED
+(
+  PDLogBase::CL_ASTAIRE_ID + 3,
+  LOG_ERR,
+  "Rogers ended - Termination signal received - terminating.",
+  "Rogers has been terminated by Monit or has exited.",
+  "The Rogers service is not longer available.",
+    "(1). This occurs normally when Rogers is stopped. "
+    "(2). If Astaire fails to respond to monit queries in a timely manner, monit restarts the application."
```
Comment (sebrexmetaswitch) [inline include/astaire_pd_definitions.hpp]: This should be Rogers not Astaire

File: rogers.root/usr/share/clearwater/infrastructure/alarms/rogers_alarms.json [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            "index": 5550,
+            "cause": "SOFTWARE_ERROR",
+            "name": "ROGERS_PROCESS_FAIL",
+            "levels": [
+                {
+                    "severity": "CLEARED",
+                    "details": "The Rogers process has been restored to normal operation.",
+                    "description": "Rogers: Process failure cleared.",
+                    "cause": "The Rogers process has been restored to normal operation. The previously issued alarm has been cleared.",
+                    "effect": "Memcached synchronization for cluster nodes is fully operational.",
```
Comment (sebrexmetaswitch) [inline rogers.root/usr/share/clearwater/infrastructure/alarms/rogers_alarms.json]: See comment just below about the working of the effect

File: rogers.root/usr/share/clearwater/infrastructure/alarms/rogers_alarms.json [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                    "action": "No action."
+                },
+                {
+                    "severity": "CRITICAL",
+                    "details": "Monit has detected that the Rogers process has failed. A restart will automatically be attempted. If this alarm does not clear, the Rogers process may have been stopped or an unrecoverable failure may have occurred.",
+                    "description": "Rogers: Process failure.",
+                    "cause": "Monit has detected that the Rogers process has failed. A restart will automatically be attempted.",
+                    "effect": "Synchronization of the Memcached database will be delayed until the process recovers.",
```
Comment (sebrexmetaswitch) [inline rogers.root/usr/share/clearwater/infrastructure/alarms/rogers_alarms.json]: The effect here is nothing to do with synchronization; rather the effect is that the memcached proxy service is not available, and any service attempting to connect via this proxy will fail to do so

File: rogers.root/usr/share/clearwater/infrastructure/monit_uptime/check-rogers-uptime [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
+
+# @file check-rogers-uptime
+#
+# Copyright (C) Metaswitch Networks 2016
```
Comment (sebrexmetaswitch) [inline rogers.root/usr/share/clearwater/infrastructure/monit_uptime/check-rogers-uptime]: 2017

File: rogers.root/usr/share/clearwater/infrastructure/monit_uptime/check-rogers-uptime [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+# If license terms are provided to you in a COPYING file in the root directory
+# of the source code repository by which you are accessing this code, then
+# the license outlined in that COPYING file applies to your use.
+# Otherwise no rights are granted except for those provided to you by
+# Metaswitch Networks in a separate written agreement.
+
+# Monit 5.8.1 does not support passing arguments to check program scripts.
+# check-uptime provides common uptime-checking code. This wrapper script
+# uses it, and can be called with no arguments.
+/usr/share/clearwater/bin/check-uptime /var/run/rogers/rogers.pid monit 5500.1
```
Comment (sebrexmetaswitch) [inline rogers.root/usr/share/clearwater/infrastructure/monit_uptime/check-rogers-uptime]: I think you want alarm id 5550.1 (as that's the Rogers equivalent to Astaire's 5500.1)

File: rogers.root/usr/share/clearwater/infrastructure/scripts/restart/rogers_restart [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+# @file rogers_restart
+#
+# Copyright (C) Metaswitch Networks 2015
```
Comment (sebrexmetaswitch) [inline rogers.root/usr/share/clearwater/infrastructure/scripts/restart/rogers_restart]: 2017

File: rogers.root/usr/share/clearwater/infrastructure/scripts/rogers.monit [hunk 9]
```diff
@@ -0,0 +1,39 @@
+#!/bin/sh
+
+# @file rogers.monit
+#
+# Copyright (C) Metaswitch Networks 2016
```
Comment (sebrexmetaswitch) [inline rogers.root/usr/share/clearwater/infrastructure/scripts/rogers.monit]: 2017

File: src/proxy_main.cpp [hunk 10]
```diff
@@ -0,0 +1,307 @@
+/**
+ * @file main.cpp - Proxy entry point
+ *
+ * Copyright (C) Metaswitch Networks 2016
```
Comment (sebrexmetaswitch) [inline src/proxy_main.cpp]: 2017

File: src/proxy_main.cpp [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+      exit(2);
+    }
+  }
+
+  return 0;
+}
+
+static sem_t term_sem;
+
+// Signal handler that triggers astaire termination.
```
Comment (sebrexmetaswitch) [inline src/proxy_main.cpp]: rogers, not astaire

File: src/proxy_main.cpp [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+  if (!view_cfg->read_config(dummy_cfg))
+  {
+    TRC_ERROR("Cluster view config is invalid. Exiting");
+    return 3;
+  }
+
+  // Create communication monitor for memcached
+  CommunicationMonitor* memcached_comm_monitor = new CommunicationMonitor(new Alarm(alarm_manager,
+                                                                                    "astaire",
+                                                                                    AlarmDef::ASTAIRE_MEMCACHED_COMM_ERROR,
```
Comment (sebrexmetaswitch) [inline src/proxy_main.cpp]: I think this should be `ROGERS_MEMCACHED_COMM_ERROR`

File: src/proxy_main.cpp [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+                                                                                    AlarmDef::CRITICAL),
+                                                                          "Rogers",
+                                                                          "Memcached");
+  // Create vbucket alarm
+  Alarm* vbucket_alarm = new Alarm(alarm_manager,
+                                   "astaire",
+                                   AlarmDef::ASTAIRE_VBUCKET_ERROR,
```
Comment (sebrexmetaswitch) [inline src/proxy_main.cpp]: Likewise, this should be `ROGERS_VBUCKET_ERROR`

File: include/astaire_pd_definitions.hpp [hunk 14]
```diff
@@ -105,4 +105,47 @@ const static PDLog CL_ASTAIRE_COMPLETE_RESYNC
...
+  LOG_INFO,
+  "Rogers started.",
+  "The Rogers application is starting.",
+  "Normal.",
+  "None."
+);
+
+const static PDLog CL_ROGERS_ENDED
+(
+  PDLogBase::CL_ASTAIRE_ID + 3,
```
Comment (sebrexmetaswitch) [inline include/astaire_pd_definitions.hpp]: This ID is copied from the Astaire equivalent above - it must be different

File: include/astaire_pd_definitions.hpp [hunk 15]
```diff
@@ -105,4 +105,47 @@ const static PDLog CL_ASTAIRE_COMPLETE_RESYNC
...
+  "Rogers ended - Termination signal received - terminating.",
+  "Rogers has been terminated by Monit or has exited.",
+  "The Rogers service is not longer available.",
+    "(1). This occurs normally when Rogers is stopped. "
+    "(2). If Astaire fails to respond to monit queries in a timely manner, monit restarts the application."
+);
+
+const static PDLog1<const char*> CL_ROGERS_TERMINATED
+(
+  PDLogBase::CL_ASTAIRE_ID + 4,
```
Comment (sebrexmetaswitch) [inline include/astaire_pd_definitions.hpp]: Again, this ID is copied from the Astaire equivalent above - it must be different

File: Makefile [hunk 16]
```diff
@@ -11,7 +11,7 @@ MODULE_DIR := ${ROOT}/modules
 
 DEB_COMPONENT := astaire
```
Comment (sebrexmetaswitch) [inline Makefile]: Does it matter that our DEB_COMPONENT only refers to Astaire? I don't think so, as this just seems to be used in the changelog (https://github.com/Metaswitch/clearwater-build-infra/blob/master/cw-deb.mk#L68) but worth considering

**Comment by sebrexmetaswitch** [comment]
I've also sent you an email with some thoughts on where we need to document this

**Comment by sebrexmetaswitch** [comment]
Also, Mat points out a bunch of places where we refer to the "memcachedstore" (e.g. in the `memcached_backend.hpp` and `.cpp` think they're called `memcachedstore.*`) and it'd be nice to tidy them up as part of this

File: Makefile (hunk 16 above)
Comment (richardwhiuk) [reply Makefile]: No it doesn't matter - this is the name of the source package, which is just named Astaire.

For similar reasons, we don't need a rogers.mk file - we have a single source directory producing a single source package, which produces multiple binary packages.

**Pull request merged**